okhttp/src/jvmTest/kotlin/okhttp3/TrailersTest.kt

```diff
@@ -0,0 +1,465 @@
+      assertThat(response.trailers()).isEqualTo(headersOf("t1", "v2"))
+    }
+  }
+
+  @Test
+  fun delayBeforeTrailersHttp1() {
+    delayBeforeTrailers(Protocol.HTTP_1_1)
+  }
+
+  @Test
+  fun delayBeforeTrailersHttp2() {
+    trailers(Protocol.HTTP_2)
+  }
+
+  /** Confirm the client will block if necessary to consume trailers. */
+  private fun delayBeforeTrailers(protocol: Protocol) {
+    enableProtocol(protocol)
+
+    server.enqueue(
+      MockResponse
+        .Builder()
+        .trailers(headersOf("t1", "v2"))
+        .body(protocol, "Hello")
+        .trailersDelay(500, TimeUnit.MILLISECONDS)
+        .build(),
+    )
+
+    val call = client.newCall(Request(server.url("/")))
+    call.execute().use { response ->
+      val source = response.body.source()
+      assertThat(source.readUtf8(5)).isEqualTo("Hello")
+      val trailersDelay =
+        measureTime {
+          val trailers = response.trailers()
+          assertThat(trailers).isEqualTo(headersOf("t1", "v2"))
+        }
+      assertThat(trailersDelay).isGreaterThan(250.milliseconds)
+    }
+  }
+
+  @Test
+  fun disconnectBeforeTrailersHttp1() {
+    disconnectBeforeTrailers(Protocol.HTTP_1_1)
+  }
+
+  @Test
+  fun disconnectBeforeTrailersHttp2() {
+    disconnectBeforeTrailers(Protocol.HTTP_2)
+  }
+
+  /** Confirm we can get an [IOException] reading trailers. */
+  private fun disconnectBeforeTrailers(protocol: Protocol) {
+    enableProtocol(protocol)
+
+    server.enqueue(
+      MockResponse
+        .Builder()
+        .trailers(headersOf("t1", "v2"))
+        .body(protocol, "Hello")
+        .socketPolicy(SocketPolicy.DisconnectDuringResponseBody)
+        .build(),
+    )
+
+    val call = client.newCall(Request(server.url("/")))
+    call.execute().use { response ->
+      assertFailsWith<IOException> {
+        response.trailers()
+      }
+    }
+  }
+
+  @Test
+  fun cannotReadTrailersAfterEarlyResponseCloseHttp1() {
+    cannotReadTrailersAfterEarlyResponseClose(Protocol.HTTP_1_1)
+  }
+
+  @Test
+  fun cannotReadTrailersAfterEarlyResponseCloseHttp2() {
+    cannotReadTrailersAfterEarlyResponseClose(Protocol.HTTP_2)
+  }
+
+  private fun cannotReadTrailersAfterEarlyResponseClose(protocol: Protocol) {
+    enableProtocol(protocol)
+
+    server.enqueue(
+      MockResponse
+        .Builder()
+        .trailers(headersOf("t1", "v2"))
+        .bodyDelay(1, TimeUnit.SECONDS)
+        .body(protocol, "Hello")
+        .build(),
+    )
+
+    val call = client.newCall(Request(server.url("/")))
+    call.execute().use { response ->
+      response.close()
+      assertFailsWith<IOException> {
+        response.trailers()
+      }
+    }
+  }
+
+  @Test
+  fun readTrailersAfterEarlyEofAndCloseHttp1() {
+    readTrailersAfterEarlyEofAndClose(Protocol.HTTP_1_1)
+  }
+
+  @Test
+  fun readTrailersAfterEarlyEofAndCloseHttp2() {
+    readTrailersAfterEarlyEofAndClose(Protocol.HTTP_2)
+  }
+
+  private fun readTrailersAfterEarlyEofAndClose(protocol: Protocol) {
```

Comment:
Maybe missing something obvious, but what makes this early EOF?

Reply:
This test name is trying too hard to differentiate itself from its neighbours. The caller is reading the end of the stream and closing it, as opposed to calling close() _after_ reading the trailers.